src/Rules/RuleRunner.cs

```diff
@@ -44,35 +44,43 @@ public IEnumerable<RunResult> RunAll(IA11yElement element)
 
             foreach (var rule in _provider.All)
             {
-                var result = RunRule(rule, element);
+                var result = RunRuleWithRetries(rule, element);
                 results.Add(result);
             } // for all rules
 
             return results;
         }
 
-        private static RunResult RunRule(IRule rule, IA11yElement element)
+        private static RunResult RunRuleWithRetries(IRule rule, IA11yElement element)
         {
-            try
+            int remainingAttempts = 3;
+
+            while (true)
             {
-                return RunRuleWorker(rule, element);
-            }
+                try
+                {
+                    return RunRuleWorker(rule, element);
+                }
 #pragma warning disable CA1031
-            catch (Exception ex)
+                catch (Exception ex)
 #pragma warning restore CA1031
-            {
-                ex.ReportException();
-                if (System.Diagnostics.Debugger.IsLogging())
-                    System.Diagnostics.Debug.WriteLine(ex.ToString());
-
-                return new RunResult
                 {
-                    RuleInfo = rule.Info,
-                    element = element,
-                    EvaluationCode = EvaluationCode.RuleExecutionError,
-                    ErrorMessage = ex.ToString()
-                };
-            } // catch
+                    ex.ReportException();
+                    if (System.Diagnostics.Debugger.IsLogging())
+                        System.Diagnostics.Debug.WriteLine(ex.ToString());
+
+                    if (--remainingAttempts == 0)
+                    {
+                        return new RunResult
+                        {
+                            RuleInfo = rule.Info,
+                            element = element,
+                            EvaluationCode = EvaluationCode.RuleExecutionError,
+                            ErrorMessage = ex.ToString()
+                        };
+                    }
+                } // catch
+            }
```

Comment:
Could a `for` loop be sufficient for this? Also although it be a VERY niche case, what if the different runs return different exceptions? I think it would be fine to just pick one but none the less maybe other folks have different opinions on this?

Reply:
Thanks, @chingucoding. I explicitly didn't want a for loop here so that we only had 2 exit paths--either we succeeded or we threw. I considered aggregating the exceptions, but when I followed how we use RunResult.ErrorMessage, we just throw it away further upstream. Each separate exception would get reported in line 68, so the telemetry would be able to view them all.

Reply:
Ah I see, thank you for the clarification @DaveTryon !